File: deepspace/SpecificDamage.java
/*
 * @author: Jesús García León
 * @file: 
 */
package deepspace;

import java.util.ArrayList;
import java.util.Iterator;

/**
 *
 * @author usuario
 */
public class SpecificDamage extends Damage{
    ArrayList<WeaponType> weapons;
    
    SpecificDamage(ArrayList<WeaponType> wl, int s){
        super(s);
        weapons = new ArrayList<>(wl);
    }
    
    @Override
    public SpecificDamageToUI getUIversion() {
        return new SpecificDamageToUI(this);
    }

    @Override
    public Damage adjust(ArrayList<Weapon> w, ArrayList<ShieldBooster> s) {
        int minNShield = Math.min(nShields, s.size());
        
            ArrayList<WeaponType> typesCopy = new ArrayList<>();
            ArrayList<Weapon> weaponsCopy = new ArrayList<>();
            ArrayList<WeaponType> adjustTypes = new ArrayList<>();
            
            for(int i=0; i< weapons.size(); i++){
                typesCopy.add(weapons.get(i));
            }
            
            for(int i=0; i<w.size(); i++){
                weaponsCopy.add(w.get(i));
            }
            
            Iterator<WeaponType> it = typesCopy.iterator();
            int index;
            
            while(it.hasNext()){
               WeaponType type = it.next();
               
               index = arrayContainsType(weaponsCopy,type);
               if(index != -1){
                   weaponsCopy.remove(index);
                   adjustTypes.add(type);
               }
               it.remove();
            }
            
            return new SpecificDamage(adjustTypes,minNShield);
    }

    @Override
    public void discardWeapon(Weapon w) {
        weapons.remove(weapons.indexOf(w.getType()));
    }

    @Override
    public boolean hasNoEffect() {
       return ((nShields == 0) && (weapons.isEmpty()));
    }

    public ArrayList<WeaponType> getWeapons() {
        return weapons;
    }

    Damage copy(SpecificDamage d) {
        super.copy(d);
        for(WeaponType w : d.weapons){
            weapons.add(w);
        }
        return this;
    }
    
    @Override
    public String toString(){
        String aux;
        aux = super.toString();
        aux += "\nWeapons:\n";
        if(weapons == null || weapons.isEmpty()){
            aux += "Ninguno\n";
        }
        else{
            aux += weapons.toString() + "\n";
        }
        return aux;
    }
}

File: DeepspaceRuby/lib/SpecificDamage.rb
#encoding:utf-8

require_relative 'SpecificDamageToUI'
require_relative 'Damage'

module Deepspace
    class SpecificDamage < Damage

        def initialize(w,nS)
            super(nS)
            @weapons=w.dup
        end

        def copy
            return SpecificDamage.new(@weapons,@nShields)
        end

        private
        def arrayContainsType(w,t)
            index = 0
            error = -1
            w.each { |wt|
                if wt.type == t
                    return index                  
                else
                    index += 1
                end
            }
            return error
        end

        public

        def getUIversion
            return SpecificDamageToUI.new(self)
        end

        def discardWeapon(w)
            index = @weapons.index(w.type)
            if index != nil
              @weapons.delete_at(index)
            end
        end

        def hasNoEffect
            return @weapons.empty? && super
        end

        def weapons
            @weapons
        end

        def adjust(w,s)
            minNShields = [@nShields,s.length].min
            vectorWeaponType = []
            vectorWeaponCopy = []

            @weapons.each{ |weapontype|
                vectorWeaponType.push(weapontype)
            }

            w.each{ |weapon|
                vectorWeaponCopy.push(weapon)
            }

            adjustType = Array.new

            @weapons.each{|weapontype|
                index = arrayContainsType(vectorWeaponCopy,weapontype)
                if index != -1
                    adjustType.push(weapontype)
                    vectorWeaponCopy.delete_at(index)
                end
            }
            return SpecificDamage.new(adjustType, minNShields)
        end

        def to_s
            string = "nShields: #{@nShields} "
            string += "Weapons: ["
            if @weapons == nil
                string += "none]\n"
            else
                for i in (0...@weapons.length-1) do
                    w = @weapons[i]
                    string += w.to_s
                    string += ","
                end
                w = @weapons[@weapons.length-1]
                string += w.to_s
                string += "]\n"
            end
            return string
        end
    end
end